Add gRPC testing guide (#1846)
grafana/k6-docs: docs/sources/k6/next/testing-guides/performance-testing-grpc-services.md

---
title: Performance testing gRPC services
description: Learn the basics of gRPC and how to write a gRPC performance test in k6.
weight: 350
---

# Performance testing gRPC services

In this guide, you'll learn the basics of gRPC and how to write a gRPC performance test using k6.

## Before you begin

> - [Install k6 v0.29](https://grafana.com/docs/k6/<K6_VERSION>/set-up/install-k6/) or higher

Reviewer: Do we end lists with a dot?

```suggestion
- [Install k6 v0.29](https://grafana.com/docs/k6/<K6_VERSION>/set-up/install-k6/) or higher.
```

Author: Yes, we should! Thanks for catching that.

For this guide, you can use the `quickpizza.proto` definition available for download on the [QuickPizza demo application GitHub repository](https://raw.githubusercontent.com/grafana/quickpizza/refs/heads/main/proto/quickpizza.proto). For details on how to build your own gRPC proto definition, refer to [the official gRPC docs](https://grpc.io/docs/what-is-grpc/core-concepts/).

```protobuf
syntax = "proto3";
option go_package = "pkg/grpc/quickpizza";
package quickpizza;

service GRPC {
    rpc Status(StatusRequest) returns (StatusResponse) {}
    rpc RatePizza(PizzaRatingRequest) returns (PizzaRatingResponse) {}
}

message StatusRequest {
}

message StatusResponse {
    bool ready = 1;
}

message PizzaRatingRequest {
    repeated string ingredients = 1;
    string dough = 2;
}

message PizzaRatingResponse {
    int32 stars_rating = 1;
}
```

## Write a gRPC performance test with k6

Starting with k6 v0.29.0, you can use a built-in module for gRPC communication. You can find more details for all the methods available on the [k6/net/grpc module documentation](https://grafana.com/docs/k6/<K6_VERSION>/javascript-api/k6-net-grpc/).

### Create the test

> The gRPC module is a separate module, available from your test script as `k6/net/grpc`. Before you can use it, you first have to create an instance of the client. Instantiating the client, as well as the `.load` operation, is only available during test initialization, that is, directly in the global scope.

Reviewer: Here, we say _"`.load` operation"_, but below, we say _"`.load()` function"_. Should we keep them consistent, like `.load()` function?

I also wonder if using a dot prefix is a common practice in our docs. I'd omit the dot for simplicity (i.e., `load()` function). Or, it can be `Client.load()` if we want to keep it verbose and explicit.

Author: That's a good point, I think we can just refer to it as `load()` function.

> The gRPC module is a separate module, available from your test script as `k6/net/grpc`. Before you can use it, you first have to create an instance of the client. Instantiating the client, as well as the `.load` operation, is only available during test initialization, that is, directly in the global scope.

Author: ```suggestion
The gRPC module is a separate module, available from your test script as `k6/net/grpc`. Before you can use it, you first have to create an instance of the client. Instantiating the client, as well as the `load()` function, is only available during test initialization, that is, directly in the global scope.
```

<!-- md-k6:skip -->

```javascript
import grpc from 'k6/net/grpc';
import { check, sleep } from 'k6';

const client = new grpc.Client();
```

> Next, load a `.proto` definition applicable for the system under test. For the purpose of this article, you can use [QuickPizza](grpc-quickpizza.grafana.com:443). The `.load()` function takes two arguments, the first one being an array of paths to search for proto files, and the second being the name of the file to load.

Reviewer: Clicking on the QuickPizza link won't work for every user. Should we add a notice?

> Next, load a `.proto` definition applicable for the system under test. For the purpose of this article, you can use [QuickPizza](grpc-quickpizza.grafana.com:443). The `.load()` function takes two arguments, the first one being an array of paths to search for proto files, and the second being the name of the file to load.

Author: ```suggestion
Next, load a `.proto` definition applicable to the system under test. For the purpose of this article, you can use [QuickPizza](grpc-quickpizza.grafana.com:443).

{{< admonition type="note" >}}

The QuickPizza gRPC service URL grpc-quickpizza.grafana.com:443 returns a 464 HTTP status code if you try to access it in your browser. However, you can still use it in your k6 test scripts, and the gRPC functionality works correctly.

{{< /admonition >}}

The `.load()` function takes two arguments, the first one being an array of paths to search for proto files, and the second being the name of the file to load.
```

Maybe something like this @inancgumus?

Reviewer: Perfect 🎉

1. The script then continues by creating the object you want to send to the remote procedure you're invoking. In the case of `RatePizza`, the object includes the ingredients for the pizza, and the type of dough.
1. Next, it invokes the remote procedure, using the syntax `<package>.<service>/<procedure>`, as described in the proto file. This call is made synchronously, with a default timeout of 60000 ms (60 seconds). To change the timeout, add the key `timeout` to the config object of `.connect()` with the duration as the value, for instance `'2s'` for 2 seconds.
1. After k6 receives a response from the server, the script checks to make sure the procedure executed successfully. The gRPC module includes constants for this comparison which are listed here. Comparing the response status with `grpc.StatusOK`, which is `200 OK` just like for HTTP/1.1 communication, ensures the call was completed successfully.
1. The script then logs the message in the response, closes the client connection, and sleeps for a second.

### Run the test

> The test can be executed just like any other test by using the k6 run command:

Reviewer: Active voice?

```suggestion
You can execute the test just like any other test by using the `k6 run` command:
```

     ✓ status is OK

     checks...............: 100.00% 1 out of 1
     data_received........: 4.1 kB  3.6 kB/s
     data_sent............: 762 B   656 B/s
     grpc_req_duration....: avg=33.75ms min=33.75ms med=33.75ms max=33.75ms p(90)=33.75ms p(95)=33.75ms
     iteration_duration...: avg=1.16s   min=1.16s   med=1.16s   max=1.16s   p(90)=1.16s   p(95)=1.16s
     iterations...........: 1       0.860427/s
     vus..................: 1       min=1      max=1
     vus_max..............: 1       min=1      max=1

running (00m01.2s), 0/1 VUs, 1 complete and 0 interrupted iterations
default ✓ [======================================] 1 VUs  00m01.2s/10m0s  1/1 iters, 1 per VU
```

From the output, you can check that the script is working and that the server indeed responds with a rating for the type of pizza supplied in the request body. You can also see that the `check` was successful, meaning the server responded with `200 OK`.

## Summary

In this article, you've gone through some of the fundamentals of gRPC and how it works. You also had a look at the k6 gRPC client and created a working test script demonstrating its functionality.

## Additional resources

> - [k6 gRPC Module API](https://grafana.com/docs/k6/<K6_VERSION>/javascript-api/k6-net-grpc/)
> - [k6 QuickPizza: A demo service for gRPC](grpc-quickpizza.grafana.com:443)

Reviewer: With dots:

```suggestion
- [k6 gRPC Module API](https://grafana.com/docs/k6/<K6_VERSION>/javascript-api/k6-net-grpc/).
- [k6 QuickPizza: A demo service for gRPC](grpc-quickpizza.grafana.com:443).
```
